crates/core/tedge/src/cli/certificate/error.rs

```diff
@@ -106,69 +106,3 @@ impl CertError {
         }
     }
 }
-
-// Our source of error here is quite deep into the dependencies and we need to dig through that to get to our certificates validator errors which are Box<&dyn Error> through 3-4 levels
-// source: hyper::Error(
-//     Connect,
-//     Custom {
-//         kind: Other,
-//         error: Custom {
-//             kind: InvalidData,
-//             error: InvalidCertificateData(
-//                 ..., // This is where we need to get
-//             ),
-//         },
-//     },
-// )
-// At the last layer we have the InvalidCertificateData error which is a Box<&dyn Error> derived from WebpkiError not included anymore, just as a String
-// This chain may break if underlying crates change.
-pub(crate) fn get_webpki_error_from_reqwest(err: reqwest::Error) -> CertError {
-    if let Some(rustls::Error::InvalidCertificateData(inner)) = err
-        // get `hyper::Error::Connect`
-        .source()
-        .and_then(|err| err.source())
-        // From here the errors are converted from std::io::Error.
-        // `Custom` type is `std::io::Error`; this is our first `Custom`.
-        .and_then(|custom_error| custom_error.downcast_ref::<std::io::Error>())
-        .and_then(|custom_error| custom_error.get_ref())
-        // This is our second `Custom`.
-        .and_then(|custom_error2| custom_error2.downcast_ref::<std::io::Error>())
-        .and_then(|custom_error2| custom_error2.get_ref())
-        // Get final error type from `rustls::Error`.
-        .and_then(|rustls_error| rustls_error.downcast_ref::<rustls::Error>())
-    {
-        match inner {
```

Comment:
This part must be move to the `certificate` crate, so this crate will no more depend on `rustls`.

Reply:
refactored.

Reply:
To fully remove the dependence to `rustls`, one needs to change the type signature of the `translate_error` function. It must take a `rustls::Error::InvalidCertificateData` parameter and not a `String`.

```Rust
if let Some(tls_error) = err.source() {
     CertError::CertificateError(translate_error(tls_error))
} else {
     CertError::ReqwestError(err)
}
```

Reply:
The `err.source()` does not emit `rustls::Error::InvalidCertificateData`, but `&dyn std::error::Error`. 
Am I missing something here?

Reply:
> Am I missing something here?

No, I was missing something.  However, I think we can remove the dependency to rustls by moving the last `downcast_ref` to the `certificate` crate:

```
pub fn translate_error(err: &dyn std::error::Error) -> Option<CertificateError> {
   if let Some(rustls::Error::InvalidCertificateData(inner)) = err {
      Some( ... )
   } else {
     None
   }
}
```

So one can write:
```
if let Some(tls_error) = err
        // get `hyper::Error::Connect`
        .source()
        // From here the errors are converted from std::io::Error.
        // `Custom` type is `std::io::Error`; this is our first `Custom`.
        .and_then(|custom_error| custom_error.downcast_ref::<std::io::Error>())
        .and_then(|custom_error| custom_error.get_ref())
        // This is our second `Custom`.
        .and_then(|custom_error2| custom_error2.downcast_ref::<std::io::Error>())
        .and_then(|custom_error2| custom_error2.get_ref())
        // Get final error type from `rustls::Error`.
        .and_then(translate_error)
{
     CertError::CertificateError(tls_error)
} else {
     CertError::ReqwestError(err)
}
```